src/thread.rs

```diff
@@ -104,10 +114,182 @@ enum ThreadJoinStatus {
     Joined,
 }
 
+/// Set of sync objects that can have properties queried.
+/// The futex is not included since it can only signal
+/// and awake values.
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+enum SyncObject {
+    /// Can query lock state.
+    Mutex(MutexId),
+    /// Can query lock state.
+    RwLock(RwLockId),
+    /// Can query the if awaited.
+    Condvar(CondvarId),
+
+}
+
+enum YieldRecordState {
+
+    /// The thread has made progress and so the recording of
+    /// states has terminated.
+    /// A thread is considered to have made progress performs some
+    /// action that is visible to another thread.
+    /// Examples include a release-store, mutex-unlock,
+    /// thread-termination and release-fence.
+    MadeProgress,
+
+    /// The thread is currently recording the variable watch
+    /// state for the yield operation, this watch operation
+    /// can occur once of multiple times to validate the yield
+    /// is correct.
+    /// Multiple iterations can prevent reporting live-lock
+    /// in yield loops with finite iterations
```

Comment:
I find this description confusing. Maybe start a bit higher-level: after a `yield`, until the next yield or until the thread makes progress, we *record* which variables the thread "watches" / depends on. This is used to wake up this thread again when some other thread "wakes" one of these "watcherd" variables.

Is that even a correct summary of what happens?